File: src/visual/src/obstacle_pcl.cpp
#include <ros/ros.h>
#include <pcl/point_cloud.h>
#include <pcl_conversions/pcl_conversions.h>
#include <sensor_msgs/PointCloud2.h>
using namespace std;

struct Point2D {
    float x, y;
};

bool pointCompare(const Point2D &p1, const Point2D &p2) {
    return p1.y < p2.y;
}

void fillPolygon(const std::vector<Point2D> &vertices, pcl::PointCloud<pcl::PointXYZ>::Ptr cloud, float resolution = 0.01, float height = 0.5) {
    if (vertices.size() < 3) return;

    auto [minIt, maxIt] = std::minmax_element(vertices.begin(), vertices.end(), pointCompare);
    float minY = minIt->y;
    float maxY = maxIt->y;

    for (float y = minY; y <= maxY; y += resolution) {
        std::vector<float> intersections;
        for (size_t i = 0; i < vertices.size(); ++i) {
            size_t j = (i + 1) % vertices.size();
            float y1 = vertices[i].y, y2 = vertices[j].y;
            float x1 = vertices[i].x, x2 = vertices[j].x;

            if ((y1 < y && y2 >= y) || (y2 < y && y1 >= y)) {
                float x = x1 + (y - y1) / (y2 - y1) * (x2 - x1);
                intersections.push_back(x);
            }
        }

        std::sort(intersections.begin(), intersections.end());

        for (size_t k = 0; k < intersections.size(); k += 2) {
            if (k + 1 >= intersections.size()) break;
            float xStart = intersections[k];
            float xEnd = intersections[k + 1];

            for (float x = xStart; x <= xEnd; x += resolution) {
                pcl::PointXYZ point;
                point.x = x;
                point.y = y;
                for (float z = 0.0; z <= height; z+= resolution){
                    point.z = z;
                    cloud->points.push_back(point);
                }
            }
        }
    }

    cloud->header.frame_id = "map";
    cloud->width = cloud->points.size();
    cloud->height = 1;
}

int main (int argc, char** argv){

    ros::init(argc,argv,"obstacle_pcl_vis");
    ros::NodeHandle nh;

    ros::Publisher pcl_pub = nh.advertise<sensor_msgs::PointCloud2>("ob_pcl_vis",1);
    pcl::PointCloud<pcl::PointXYZ>::Ptr obstacle_ptr(new pcl::PointCloud<pcl::PointXYZ>);
    
    // construct cloud point data
    float height = 0.4;
    float resolution = 0.05;
    std::vector<Point2D> empty;
    std::vector<Point2D> vertices;

    vertices = {
        {1.0, 1.0}, {2.0, 1.0}, {2.0,2.0}, {1.0,2.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);
    
    // generate a new obstacle point cloud 
    vertices.swap(empty);
    vertices = {
        {3.0, 1.0}, {4.0, 1.0}, {4.0,3.0}, {3.0,3.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);
    
    vertices.swap(empty);
    vertices = {
        {1.0, 4.0}, {3.0, 4.0}, {3.0,5.0}, {1.0,5.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {4.0, 4.0}, {5.0, 4.0}, {5.0,8.0}, {4.0,8.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {1.0, 8.0}, {2.0, 8.0}, {2.0,9.0}, {1.0,9.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {6.0, 3.0}, {8.0, 3.0}, {8.0,5.0}, {6.0,5.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {6.0, 1.0}, {7.0, 1.0}, {7.0,2.0}, {6.0,2.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {7.0, 8.0}, {8.0, 8.0}, {8.0,9.0}, {7.0,9.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {7.0, 6.0}, {9.0, 6.0}, {9.0,7.0}, {7.0,7.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {0.0, 6.0}, {1.0, 6.0}, {1.0,7.0}, {0.0,7.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {2.0, 6.0}, {3.0, 6.0}, {3.0,7.0}, {2.0,7.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    vertices.swap(empty);
    vertices = {
        {4.0, 2.0}, {5.0, 2.0}, {5.0, 3.0}, {4.0, 3.0}
    };
    fillPolygon(vertices, obstacle_ptr, resolution, height);

    // convert to ROS message
    sensor_msgs::PointCloud2 output;
    pcl::toROSMsg(*obstacle_ptr,output);

    ros::Rate rate(1);
    while(ros::ok()){
        pcl_pub.publish(output);
        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}

File: src/visual/scripts/obstacle_pcl.py
################################################################
# 
# Author: Hauser Dong
# From Peking university
# Last update: 2025.03.03
# 
###############################################################

import rospy
import pcl
import sys
import sensor_msgs.point_cloud2 as pc2
from sensor_msgs.msg import PointCloud2, PointField
from std_msgs.msg import Header
import numpy as np
import os

current_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.append(os.path.join(current_dir, '../../local_planner/planner/scripts'))
import SET
from geometry import *

class Point2D:
    def __init__(self, x, y):
        self.x = x
        self.y = y


def point_compare(p1, p2):
    return p1.y < p2.y


def fill_polygon(vertices, cloud, resolution=0.01, height=0.5):
    if len(vertices) < 3:
        return

    min_y = min(vertices, key=lambda p: p.y).y
    max_y = max(vertices, key=lambda p: p.y).y

    for y in np.arange(min_y, max_y + resolution, resolution):
        intersections = []
        for i in range(len(vertices)):
            j = (i + 1) % len(vertices)
            x1, y1 = vertices[i].x, vertices[i].y
            x2, y2 = vertices[j].x, vertices[j].y

            if (y1 < y and y2 >= y) or (y2 < y and y1 >= y):
                x = x1 + (y - y1) / (y2 - y1) * (x2 - x1)
                intersections.append(x)

        intersections.sort()

        for k in range(0, len(intersections), 2):
            if k + 1 >= len(intersections):
                break
            x_start = intersections[k]
            x_end = intersections[k + 1]

            for x in np.arange(x_start, x_end + resolution, resolution):
                for z in np.arange(0.0, height + resolution, resolution):
                    cloud.append([x, y, z])


def pcl_to_ros(cloud):
    header = Header()
    header.stamp = rospy.Time.now()
    header.frame_id = "map"
    
    fields = [PointField('x', 0, PointField.FLOAT32, 1),
              PointField('y', 4, PointField.FLOAT32, 1),
              PointField('z', 8, PointField.FLOAT32, 1)]
    
    return pc2.create_cloud(header, fields, cloud)


if __name__ == '__main__':
    rospy.init_node('obstacle_pcl_vis')

    pcl_pub = rospy.Publisher('ob_pcl_vis', PointCloud2, queue_size=1)
    pcl_pub_inflated = rospy.Publisher('ob_pcl_vis_inflated', PointCloud2, queue_size=1)
    
    resolution = 0.05
    height = 0.4
    
    obstacle_cloud = []
    obstacles_cloud_inflated = []

    obs_env_idx = rospy.get_param('obs_env_idx', default=4)
    obstacles = SET.ini_obstacle_list[obs_env_idx]

    obstacles_inflated, _ = Build_ExtensionZone(obstacles , 0.2)

    # Add polygons and convert to ROS message
    polygons = [[Point2D(vertex[0], vertex[1]) for vertex in ob.vertex_list] for ob in obstacles]
    polygons_inflated = [[Point2D(vertex[0], vertex[1]) for vertex in ob.vertex_list] for ob in obstacles_inflated]

    for polygon in polygons:
        fill_polygon(polygon, obstacle_cloud, resolution, height)
        
    for polygon in polygons_inflated:
        fill_polygon(polygon, obstacles_cloud_inflated, resolution, height)

    # Convert PCL cloud to ROS message
    output = pcl_to_ros(obstacle_cloud)
    output_inflated = pcl_to_ros(obstacles_cloud_inflated)

    rate = rospy.Rate(1)
    while not rospy.is_shutdown():
        pcl_pub.publish(output)
        pcl_pub_inflated.publish(output_inflated)
        rate.sleep()
